File: client-android/lib/src/main/kotlin/stasis/client_android/lib/collection/DefaultBackupCollector.kt
package stasis.client_android.lib.collection

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.map
import stasis.client_android.lib.api.clients.Clients
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.SourceEntity
import java.nio.file.Path

class DefaultBackupCollector(
    private val entities: List<Path>,
    private val latestMetadata: DatasetMetadata?,
    private val metadataCollector: BackupMetadataCollector,
    private val clients: Clients
) : BackupCollector {
    override fun collect(): Flow<SourceEntity> =
        collectEntityMetadata(
            entities = entities,
            latestMetadata = latestMetadata,
            clients = clients
        ).map { (entity, entityMetadata) ->
            metadataCollector.collect(entity = entity, existingMetadata = entityMetadata)
        }


    companion object {
        fun collectEntityMetadata(
            entities: List<Path>,
            latestMetadata: DatasetMetadata?,
            clients: Clients
        ): Flow<Pair<Path, EntityMetadata?>> =
            when (latestMetadata) {
                null -> entities.asFlow().map { entity -> entity to null }
                else -> entities.asFlow().map { entity -> entity to latestMetadata.collect(entity, clients) }
            }
    }
}

File: client-ios/lib/Sources/StasisClientLib/Collection/DefaultBackupCollector.swift
import Foundation

public struct DefaultBackupCollector: BackupCollector {
    private let entities: [URL]
    private let latestMetadata: DatasetMetadata?
    private let metadataCollector: any BackupMetadataCollector
    private let clients: any Clients

    public init(
        entities: [URL],
        latestMetadata: DatasetMetadata?,
        metadataCollector: any BackupMetadataCollector,
        clients: any Clients
    ) {
        self.entities = entities
        self.latestMetadata = latestMetadata
        self.metadataCollector = metadataCollector
        self.clients = clients
    }

    public func collect() -> AsyncThrowingStream<SourceEntity, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let stream = DefaultBackupCollector.collectEntityMetadata(
                        entities: entities,
                        latestMetadata: latestMetadata,
                        clients: clients
                    )
                    for try await (entity, existing) in stream {
                        let source = try await metadataCollector.collect(entity: entity, existingMetadata: existing)
                        continuation.yield(source)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    public static func collectEntityMetadata(
        entities: [URL],
        latestMetadata: DatasetMetadata?,
        clients: any Clients
    ) -> AsyncThrowingStream<(URL, EntityMetadata?), Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    for entity in entities {
                        if let latestMetadata {
                            let metadata = try await latestMetadata.collect(entity: entity, clients: clients)
                            continuation.yield((entity, metadata))
                        } else {
                            continuation.yield((entity, nil))
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}
